JustSaying.IntegrationTests/Fluent/IntegrationTestBase.cs

```diff
@@ -80,7 +80,15 @@ protected async Task WhenAsync(IServiceCollection services, Func<IMessagePublish
 
             using (var source = new CancellationTokenSource(Timeout))
             {
-                await Task.WhenAny(action(publisher, listener, source.Token), Task.Delay(Timeout, source.Token)).ConfigureAwait(false);
+                var actionTask = action(publisher, listener, source.Token);
+
+                await Task.WhenAny(actionTask, Task.Delay(Timeout, source.Token))
+                    .ConfigureAwait(false);
+
+                if (!actionTask.IsCompletedSuccessfully)
+                {
+                    throw actionTask.Exception ?? new Exception("The action failed");
+                }
```

Comment:
The other thing to add would be `source.Token.ThrowIfCancelled()` (sic).

Reply:
If the `actionTask` errors, then the exception will bubble up during `await WhenAny`. It might be worth copying the example here:
https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md
If you search for the text:
`This example cancels the timer if the operation succesfully completes.`

Reply:
```csharp
using (var source = new CancellationTokenSource(Timeout))
{
    var delayTask = Task.Delay(Timeout, source.Token);
    var resultTask = await Task.WhenAny(action(publisher, listener, source.Token), delayedTask).ConfigureAwait(false);
    
    if (resultTask == delayTask)
    {
        throw new OperationCanceledException();
    }
    else
    {
        cts.Cancel();
    }
}
```

Reply:
1) Why not just call `source.Token.ThrowIfCancelled()` if it's the task that finished?
2) If it didn't timeout, do we need to even bother canceling it?

Reply:
For 2, it just stops having timers lying about, but for tests it's probably not that important.
1, yeah that's fair.

```csharp
using (var source = new CancellationTokenSource(Timeout))
{
    var resultTask = await Task.WhenAny(action(publisher, listener, source.Token),
                                        Task.Delay(Timeout, source.Token)).ConfigureAwait(false);
    
    source.Token.ThrowIfCancelled();
    cts.Cancel();
    await resultTask;
}
```

Reply:
Doesn't the dispose block for the CTS take care of that?

Reply:
I don't think it cancels, just doing some tests

Reply:
`await resultTask` or `await actionTask` might not be safe - what if it's deadlocked or very slow. 
That's the point of the `delayTask`, right?

Reply:
confirmed, It doesn't cancel

Reply:
You wouldn't be awaiting a task that is slow or stuck because you'd be at that line only if the delay task completes, then the `source.Token.ThrowIfCancelled()` guards you from the case it was because of a timeout.

Reply:
TIL.

Reply:
Move this discussion to #487